libs/filer/workspace_files_cache.go

```diff
@@ -0,0 +1,425 @@
+func (c *cache) enqueue(ctx context.Context, e executable) error {
+	select {
+	case <-ctx.Done():
+		return ctx.Err()
+	case c.queue <- e:
+		return nil
+	default:
+		return queueFullError{e.String()}
+	}
+}
+
+func (c *cache) completeReadDirForDir(name string, dirEntry fs.DirEntry) {
+	// Add to the stat cache if not already present.
+	if _, ok := c.stat[name]; !ok {
+		e := newStatEntry(name)
+		e.markDone(dirEntry.Info())
+		c.stat[name] = e
+	}
+
+	// Queue a [ReadDir] call for the directory if not already present.
+	if _, ok := c.readDir[name]; !ok {
+		// Create a new cache entry and queue the operation.
+		e := newReadDirEntry(name)
+		err := c.enqueue(context.Background(), e)
+		if err != nil {
+			e.markError(err)
+		}
+
+		// Add the entry to the cache, even if has an error.
+		c.readDir[name] = e
+	}
+}
+
+func (c *cache) completeReadDirForFile(name string, dirEntry fs.DirEntry) {
+	// Skip if this entry is already in the cache.
+	if _, ok := c.stat[name]; ok {
+		return
+	}
+
+	// Create a new cache entry.
+	e := newStatEntry(name)
+
+	// Depending on the object type, we either have to perform a real
+	// stat call, or we can use the [fs.DirEntry] info directly.
+	switch dirEntry.(wsfsDirEntry).ObjectType {
+	case workspace.ObjectTypeNotebook:
+		// Queue a [Stat] call for the file.
+		err := c.enqueue(context.Background(), e)
+		if err != nil {
+			e.markError(err)
+		}
+	default:
+		// Use the [fs.DirEntry] info directly.
+		e.markDone(dirEntry.Info())
+	}
+
+	// Add the entry to the cache, even if has an error.
+	c.stat[name] = e
+}
+
+func (c *cache) completeReadDir(dir string, entries []fs.DirEntry) {
+	c.m.Lock()
+	defer c.m.Unlock()
+
+	for _, e := range entries {
+		name := path.Join(dir, e.Name())
+
+		if e.IsDir() {
+			c.completeReadDirForDir(name, e)
+		} else {
+			c.completeReadDirForFile(name, e)
+		}
+	}
+}
+
+// Cleanup closes the queue and waits for all goroutines to exit.
+func (c *cache) Cleanup() {
+	close(c.queue)
+	c.wg.Wait()
+}
+
+// Write passes through to the underlying Filer.
+func (c *cache) Write(ctx context.Context, name string, reader io.Reader, mode ...WriteMode) error {
+	return c.f.Write(ctx, name, reader, mode...)
+}
+
+// Read passes through to the underlying Filer.
+func (c *cache) Read(ctx context.Context, name string) (io.ReadCloser, error) {
+	return c.f.Read(ctx, name)
+}
+
+// Delete passes through to the underlying Filer.
+func (c *cache) Delete(ctx context.Context, name string, mode ...DeleteMode) error {
+	return c.f.Delete(ctx, name, mode...)
+}
+
+// Mkdir passes through to the underlying Filer.
+func (c *cache) Mkdir(ctx context.Context, name string) error {
+	return c.f.Mkdir(ctx, name)
+}
+
+// ReadDir returns the entries in a directory.
+// If the directory is already in the cache, the cached value is returned.
+func (c *cache) ReadDir(ctx context.Context, name string) ([]fs.DirEntry, error) {
+	name = path.Clean(name)
+
+	// Lock before R/W access to the cache.
+	c.m.Lock()
```

Comment:
[optional] Maybe add `defer c.m.Unlock` here instead of having the three unlock before return? I don't think it is truly needed but it's generally a good practice as it guarantees that the mutex will be unlocked in case of panic.

Reply:
Not possible because we need to unlock prior to waiting.

Reply:
Oh, you're right 👍